finality-aleph/src/data_io.rs

```diff
@@ -203,138 +448,535 @@ where
+            }
+        }
+    }
+
+    fn on_proposal_available(&mut self, proposal: &AlephProposal<B>) {
+        self.available_proposal_cache.put(proposal.clone(), ());
+        if let Some(proposal_info) = self.pending_proposals.remove(proposal) {
+            for id in proposal_info.messages {
+                self.remove_proposal_from_pending_message(proposal, id);
             }
         }
     }
 
-    fn add_pending_message(&mut self, message: Message, requirements: Vec<AlephDataFor<B>>) {
-        let message_id = self.next_message_id;
-        self.next_message_id += 1;
-        for block_data in requirements.iter() {
-            self.missing_blocks
-                .entry(*block_data)
-                .or_insert_with(MissingBlockInfo::new)
-                .messages
-                .insert(message_id);
+    // Makes an availability check for `data` and updates its status. Outputs whether the bump resulted in
+    // this proposal gaining availability. `thorough = false` uses only cached information about available
+    // blocks, `thorough = true` also uses the client.
+    fn bump_proposal(&mut self, proposal: &AlephProposal<B>, thorough: bool) -> bool {
+        let proposal_info = self.pending_proposals.get(proposal);
+        if proposal_info.is_none() {
+            // It is possible that `proposal` is not pending anymore, as it has become available as a result of
+            // an earlier bump.
+            return false;
         }
-        self.message_requirements
-            .insert(message_id, requirements.len());
-        self.pending_messages.insert(message_id, message);
+        // Some minor inefficiencies in HashMap access below because of borrow checker.
+        let old_status: ProposalStatus = proposal_info.unwrap().status.clone();
+        let new_status: ProposalStatus =
+            self.check_proposal_availability(proposal, &old_status, thorough);
+        self.pending_proposals.get_mut(proposal).unwrap().status = new_status.clone();
 
-        // Below we remove the message with id `self.next_message_id - self.config.message_id_boundary` to keep the invariant
-        // that the set of messages kept are the ones with ids in the interval
-        // [self.next_message_id - self.config.message_id_boundary, self.next_message_id-1].
-        // Note that `add_pending_message` is the only method that modifies `self.next_message_id`, this works as expected.
-        if message_id >= self.config.message_id_boundary {
-            self.forget_message(message_id - self.config.message_id_boundary)
+        use ProposalStatus::*;
+        match new_status {
+            PendingTopBlock => {
+                self.register_next_finality_trigger(proposal);
+                false
+            }
+            TopBlockImportedButIncorrectBranch => {
+                // We do nothing, this is a result of malicious behaviour. This proposal can still get through,
+                // but we wait till maintenance.
```

Comment:
We will let it through, because of liveliness. Although I agree that we might not want to let it through _yet_, didn't notice that, thanks.

Reply:
Wait, why? If `H0` gets finalized, we will never (and should not) let this proposal through.

Reply:
Let me be more precise here:
At the very moment `H0` gets finalized, this proposal does not yet satisfy the definition of available `AlephData`, so it will stay pending. At the moment a block one above `H0` gets finalized (it will be conflicting with `H1`), we will let it through, as it will be a hopeless fork, by definition (there is a comment before `DataStore` explaining this formally).

Reply:
So, I think you are right in saying that once enough blocks are finalized, we will let it trough.

Reply:
Won't a node that gets this proposal after it has finalized a `N+1`th block let it through as a hopeless fork anyway?

I read the code and I think (not sure though) that this will be let through in this case, since `is_hopeless_fork` gets called early and does not check parent consistency.

This is one of the reasons I would like all the logic about the status of a proposal based on our knowledge to be in one place and return a very specific recommendation.

Reply:
Yes, in the definition of a hopeless fork, no consistency is enforced, so after a block at height `N+1` is finalized, different than `H1`, this will get through as available.

As for moving the logic: I'm currently working on that, or at least thinking how to do that :/